From f627fddc4091666683ca115a46151cfd99c01191 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Tue, 2 Jan 2024 14:12:30 +0530
Subject: [PATCH] hostapd: validate mld_ap bss config

Currently, even if ieee80211be is disabled and mld_ap is set, hostapd tries
to access mlo related functions and structures which is wrong. At times, a
few segmentation fault is also seen.

Fix this issue by validating the bss config and if mld_ap is set with
ieee80211be unset, make mld_ap also as unset since former should not be set
when later is not.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/ap/ap_config.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -1513,6 +1513,12 @@ static int hostapd_config_check_bss(stru
 #endif /* CONFIG_FILS */
 
 #ifdef CONFIG_IEEE80211BE
+	if (full_config && !conf->ieee80211be && bss->mld_ap) {
+		bss->mld_ap = 0;
+		wpa_printf(MSG_INFO,
+			   "MLD: Disabling MLO as IEEE 802.11be is disabled");
+	}
+
 	if (full_config && !bss->disable_11be && bss->disable_11ax) {
 		bss->disable_11be = true;
 		wpa_printf(MSG_INFO,
